### DISCUSSION TITLE
How do I define grammar for `x=1,x`?

### DISCUSSION THREAD

**Discussion by panlina** [tree-sitter/tree-sitter#781] [Q&A] [answered]
I have this grammar:
```js
module.exports = grammar({
	name: 'a',
	word: $ => $.identifier,
	rules: {
		expression: $ => choice(
			$.number,
			$.identifier,
			$.expression_binary,
			$.expression_comma
		),
		expression_comma: $ => prec.right(-1, seq(
			field('name', $.identifier),
			'=',
			field('value', $.expression),
			',',
			field('body', $.expression)
		)),
		expression_binary: $ => prec.left(0, seq(
			field('left', $.expression),
			field('operator', '='),
			field('right', $.expression)
		)),
		identifier: $ => /[_a-zA-Z][_a-zA-Z0-9]*/,
		number: $ => /\d+/,
	}
});
```
I have basic number and identifier grammar, an equal comparison expression, then a "let expression", which binds a variable in an expression.

Example:
- `x`: identifier
- `x=1`: evaluates if `a` equals `1`
- `x=1,x`: binds `x` to `1`, then evaluate `x`

When I parse `x=1,x`, I get an ERROR:
```wat
(expression [0, 0] - [0, 5]
  (ERROR [0, 0] - [0, 4]
    (expression [0, 0] - [0, 3]
      (expression_binary [0, 0] - [0, 3]
        left: (expression [0, 0] - [0, 1]
          (identifier [0, 0] - [0, 1]))
        right: (expression [0, 2] - [0, 3]
          (number [0, 2] - [0, 3])))))
  (identifier [0, 4] - [0, 5]))
```
Seems it successfully matched `x=1` as an equal comparison expression, then failed to parse the remaining text, and it does not step back and try a `expression_comma` rule instead.

I'm sure I can workaround this by introducing a `let` keyword, or use `==` instead of `=` for equal comparison expression, but is it possible to get it right with the current grammar?

**Comment by maxbrunsfeld** [comment, marked as answer]
I think it's because of the `prec(-1)` around `expression_comma`. Try removing that, and checking the `Unresolved conflict` message that Tree-sitter reports. It looks like there is an ambiguity at the `,`. Maybe you can restructure the grammar to remove this ambiguity.

**Reply by panlina** [reply]
If I remove it, it outputs:
```
Unresolved conflict for symbol sequence:

  identifier  •  '='  …

Possible interpretations:

  1:  (expression  identifier)  •  '='  …
  2:  (expression_comma  identifier  •  '='  expression  ','  expression)

Possible resolutions:

  1:  Specify a higher precedence in `expression_comma` than in the other rules.
  2:  Specify a higher precedence in `expression` than in the other rules.
  3:  Specify a left or right associativity in `expression`
  4:  Add a conflict for these rules: `expression`, `expression_comma`
```
Actually I expect no ambiguity here, since in this grammar, there can't be any string that matches both interpretations, because as long as `identifier  •  '='  expression` forms an expression, the remaining `','  expression` fails.

**Reply by maxbrunsfeld** [reply]
This is what is called a "local ambiguity", or "LR(1) conflict". It means that, at a given position during parsing, it is impossible to decide how the *preceding* content should be interpreted while only looking ahead by one token. In general, checking if a grammar is truly ambiguous is an undecidable computation. But this more conservative check can be done very efficiently, and helps you to create a parser that is resource-efficient.

You can choose to have Tree-sitter try *both* interpretations at runtime by choosing option 4 (adding a conflict for `expression` and `expression_comma`. This has a small performance cost though. My advice would be to try to structure your grammar so that, when seeing the `=`, you can decide how the left hand side should be interpreted. You should only allow LR(1) conflicts as a 'last resort', if you need to parse something that fundamentally requires looking ahead.

**Reply by panlina** [reply]
Oh, appreciate your explanation. 👍. Not only answered my question, but also helped me really understand what these outputs mean. Thanks!

**Reply by panlina** [reply]
After adding conflicts, I come up with another problem.
```
module.exports = grammar({
	name: 'a',
	word: $ => $.identifier,
	rules: {
		expression: $ => choice(
			$.number,
			$.identifier,
			$.expression_binary,
			$.expression_comma
		),
		expression_comma: $ => seq(
			field('name', $.identifier),
			'=',
			field('value', $.expression),
			',',
			field('body', $.expression)
		),
		expression_binary: $ => prec.left(0, seq(
			field('left', $.expression),
			field('operator', '='),
			field('right', $.expression)
		)),
		identifier: $ => /[_a-zA-Z][_a-zA-Z0-9]*/,
		number: $ => /\d+/,
	},
	conflicts: $ => [
		[$.expression, $.expression_comma],
		[$.expression_comma, $.expression_binary],
	]
});
```
It parses `x=1,x` fine, but when parsing `x=1,x=x`, it gives `(x=1,x)=x` instead of `x=1,(x=x)`. So I tried to add `prec` for  `expression_comma` back:
```
		expression_comma: $ => prec.right(-1, seq(
			field('name', $.identifier),
			'=',
			field('value', $.expression),
			',',
			field('body', $.expression)
		)),
```
, but this made it fail to parse again.
```
(expression [0, 0] - [0, 7]
  (expression_binary [0, 0] - [0, 7]
    left: (expression [0, 0] - [0, 3]
      (expression_binary [0, 0] - [0, 3]
        left: (expression [0, 0] - [0, 1]
          (identifier [0, 0] - [0, 1]))
        right: (expression [0, 2] - [0, 3]
          (number [0, 2] - [0, 3]))))
    (ERROR [0, 3] - [0, 5]
      (identifier [0, 4] - [0, 5]))
    right: (expression [0, 6] - [0, 7]
      (identifier [0, 6] - [0, 7]))))
example-file    0 ms    (ERROR [0, 3] - [0, 5])
```
Why adding a precedence fails `conflicts`?

**Reply by panlina** [reply]
Finally I get it working by separating `expression_comma` into 2 rules:
```
		expression_comma: $ => prec.right(-1, seq(
			field('head', $.expression_comma_head),
			',',
			field('body', $.expression)
		)),a
		expression_comma_head: $ => seq(
			field('name', $.identifier),
			'=',
			field('value', $.expression),
		),
```
and rewriting `conflicts` as:
```
	conflicts: $ => [
		[$.expression_comma_head, $.expression_identifier]
	]
```
Still don't know why not doing so fails `conflicts` though.
